diff -burN --exclude=.git binutils-2.20.1/config.sub binutils-2.20.1-PS3/config.sub
--- binutils-2.20.1/config.sub	2009-08-17 01:10:29.000000000 -0300
+++ binutils-2.20.1-PS3/config.sub	2011-01-14 23:39:53.349563973 -0400
@@ -935,6 +935,10 @@
 	ppc64le-* | powerpc64little-*)
 		basic_machine=powerpc64le-`echo $basic_machine | sed 's/^[^-]*-//'`
 		;;
+	ppu)
+		basic_machine=powerpc64-unknown
+		os=-elf
+		;;
 	ps2)
 		basic_machine=i386-ibm
 		;;
diff -burN --exclude=.git binutils-2.20.1/ld/emultempl/spuelf.em binutils-2.20.1-PS3/ld/emultempl/spuelf.em
--- binutils-2.20.1/ld/emultempl/spuelf.em	2009-08-10 04:50:56.000000000 -0300
+++ binutils-2.20.1-PS3/ld/emultempl/spuelf.em	2011-01-14 23:39:53.870816093 -0400
@@ -355,7 +355,7 @@
   *fname = mktemp (*fname);
   if (*fname == NULL)
     return -1;
-  fd = open (fname, O_RDWR | O_CREAT | O_EXCL, 0600);
+  fd = open (*fname, O_RDWR | O_CREAT | O_EXCL, 0600);
 #endif
   return fd;
 }
@@ -396,7 +396,11 @@
   argv[my_argc++] = "-T";
   argv[my_argc++] = auto_overlay_file;
   argv[my_argc] = 0;
+#if defined(_WIN32)
+  execvp (argv[0], (const char* const*) argv);
+#else
   execvp (argv[0], (char *const *) argv);
+#endif
   perror (argv[0]);
   _exit (127);
 }
