diff -burN orig.binutils-2.20/config.sub binutils-2.20/config.sub
--- orig.binutils-2.20/config.sub	2009-08-17 01:10:29.000000000 -0300
+++ binutils-2.20/config.sub	2010-05-17 12:51:55.115575727 -0300
@@ -935,6 +935,10 @@
 	ppc64le-* | powerpc64little-*)
 		basic_machine=powerpc64le-`echo $basic_machine | sed 's/^[^-]*-//'`
 		;;
+	ppu)
+		basic_machine=powerpc-unknown
+		os=-elf
+		;;
 	ps2)
 		basic_machine=i386-ibm
 		;;
