diff -burN --exclude=.git binutils-2.20.1/config.sub binutils-2.20.1-PS3/config.sub
--- binutils-2.20.1/config.sub	2009-08-17 01:10:29.000000000 -0300
+++ binutils-2.20.1-PS3/config.sub	2011-03-23 21:26:08.088037883 -0300
@@ -935,6 +935,10 @@
 	ppc64le-* | powerpc64little-*)
 		basic_machine=powerpc64le-`echo $basic_machine | sed 's/^[^-]*-//'`
 		;;
+	ppu)
+		basic_machine=powerpc64-unknown
+		os=-elf
+		;;
 	ps2)
 		basic_machine=i386-ibm
 		;;
diff -burN --exclude=.git binutils-2.20.1/ld/emultempl/spuelf.em binutils-2.20.1-PS3/ld/emultempl/spuelf.em
--- binutils-2.20.1/ld/emultempl/spuelf.em	2009-08-10 04:50:56.000000000 -0300
+++ binutils-2.20.1-PS3/ld/emultempl/spuelf.em	2011-03-23 21:26:08.598037917 -0300
@@ -355,7 +355,7 @@
   *fname = mktemp (*fname);
   if (*fname == NULL)
     return -1;
-  fd = open (fname, O_RDWR | O_CREAT | O_EXCL, 0600);
+  fd = open (*fname, O_RDWR | O_CREAT | O_EXCL, 0600);
 #endif
   return fd;
 }
@@ -396,7 +396,11 @@
   argv[my_argc++] = "-T";
   argv[my_argc++] = auto_overlay_file;
   argv[my_argc] = 0;
+#if defined(_WIN32)
+  execvp (argv[0], (const char* const*) argv);
+#else
   execvp (argv[0], (char *const *) argv);
+#endif
   perror (argv[0]);
   _exit (127);
 }
diff -burN --exclude=.git binutils-2.20.1/include/elf/internal.h binutils-2.20.1-PS3/include/elf/internal.h
--- binutils-2.20.1/include/elf/internal.h	2008-11-14 15:44:09.000000000 +0100
+++ binutils-2.20.1-PS3/include/elf/internal.h	2011-05-15 16:35:15.928088040 +0200
@@ -316,6 +316,7 @@
    /* SHF_ALLOC sections must have VMAs within the segment.  Be		\
       careful about segments right at the end of memory.  */		\
    && ((sec_hdr->sh_flags & SHF_ALLOC) == 0				\
+       || ELF_SECTION_SIZE(sec_hdr,segment) == 0                        \
        || (sec_hdr->sh_addr >= segment->p_vaddr				\
 	   && (sec_hdr->sh_addr - segment->p_vaddr			\
 	       + ELF_SECTION_SIZE(sec_hdr, segment) <= segment->p_memsz))))
