diff -burN '--exclude=.git' gcc-4.6.1/gcc/config/rs6000/rs6000.c gcc-4.6.1-PS3/gcc/config/rs6000/rs6000.c
--- gcc-4.6.1/gcc/config/rs6000/rs6000.c	2011-06-20 12:12:44.000000000 -0300
+++ gcc-4.6.1-PS3/gcc/config/rs6000/rs6000.c	2011-08-15 23:44:25.464052944 -0300
@@ -28257,5 +28257,12 @@
   return x;
 }
 
+#ifdef POWERPC_CELL64LV2
+
+static bool rs6000_cell64lv2_valid_pointer_mode(enum machine_mode mode)
+{
+     return (mode == SImode || (TARGET_64BIT && mode == DImode) || mode == ptr_mode || mode == Pmode);
+}
+#endif
 
 #include "gt-rs6000.h"
