diff -burN '--exclude=.git' gdb-7.2/config.sub gdb-7.2-PS3/config.sub
--- gdb-7.2/config.sub	2010-06-01 14:53:40.000000000 -0300
+++ gdb-7.2-PS3/config.sub	2011-05-25 10:29:20.150050720 -0300
@@ -4,7 +4,7 @@
 #   2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010
 #   Free Software Foundation, Inc.
 
-timestamp='2010-05-21'
+timestamp='2010-10-20'
 
 # This file is (in principle) common to ALL GNU software.
 # The presence of a machine in this file suggests that SOME GNU software
@@ -239,6 +239,9 @@
 		basic_machine=m68k-atari
 		os=-mint
 		;;
+        -lv2)
+	        os=-elf
+		;;
 esac
 
 # Decode aliases for certain CPU-COMPANY combinations.
@@ -962,6 +965,10 @@
 	ppc64le-* | powerpc64little-*)
 		basic_machine=powerpc64le-`echo $basic_machine | sed 's/^[^-]*-//'`
 		;;
+	ppu)
+	        basic_machine=powerpc64-unknown
+		os=-linux-gnu
+		;;
 	ps2)
 		basic_machine=i386-ibm
 		;;
